arbnode/dataposter/data_poster.go

```diff
@@ -0,0 +1,386 @@
+
+func NewDataPoster[Meta any](headerReader *headerreader.HeaderReader, auth *bind.TransactOpts, config *DataPosterConfig) (*DataPoster[Meta], error) {
+	var replacementTimes []time.Duration
+	var lastReplacementTime time.Duration
+	for _, s := range strings.Split(config.ReplacementTimes, ",") {
+		t, err := time.ParseDuration(s)
+		if err != nil {
+			return nil, err
+		}
+		if t <= lastReplacementTime {
+			return nil, errors.New("replacement times must be increasing")
+		}
+		replacementTimes = append(replacementTimes, t)
+		lastReplacementTime = t
+	}
+	if len(replacementTimes) == 0 {
+		log.Warn("disabling replace-by-fee for data poster")
+	}
+	// To avoid special casing "don't replace again", replace in 10 years
+	replacementTimes = append(replacementTimes, time.Hour*24*365*10)
+	var queue QueueStorage[queuedTransaction[Meta]]
+	if config.RedisUrl == "" {
+		queue = NewSliceStorage[queuedTransaction[Meta]]()
+	} else {
+		var err error
+		queue, err = NewRedisStorage[queuedTransaction[Meta]](config.RedisUrl, "data-poster.queue")
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &DataPoster[Meta]{
+		headerReader:     headerReader,
+		client:           headerReader.Client(),
+		auth:             auth,
+		config:           config,
+		replacementTimes: replacementTimes,
+		queue:            queue,
+	}, nil
+}
+
+func (p *DataPoster[Meta]) Initialize(ctx context.Context) error {
+	nonce, err := p.client.NonceAt(ctx, p.auth.From, nil)
+	if err != nil {
+		return err
+	}
+	p.nonce = nonce
+	return nil
+}
+
+func (p *DataPoster[Meta]) From() common.Address {
+	return p.auth.From
+}
+
+func (p *DataPoster[Meta]) GetNextNonceAndMeta(ctx context.Context, getMetaAtBlock func(blockNum *big.Int) (Meta, error)) (uint64, Meta, error) {
+	var emptyMeta Meta
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	err := p.updateState(ctx)
+	if err != nil {
+		return 0, emptyMeta, err
+	}
+	lastQueueItem, err := p.queue.GetLast(ctx)
+	if err != nil {
+		return 0, emptyMeta, err
+	}
+	if lastQueueItem != nil {
+		return lastQueueItem.Data.Nonce + 1, lastQueueItem.Meta, nil
+	}
+	meta, err := getMetaAtBlock(p.lastBlock)
+	return p.nonce, meta, err
+}
+
+const minRbfIncrease arbmath.Bips = arbmath.OneInBips * 11 / 10
+
+func (p *DataPoster[Meta]) getFeeAndTipCaps(ctx context.Context, lastTipCap *big.Int, dataCreatedAt time.Time) (*big.Int, *big.Int, error) {
+	latestHeader, err := p.headerReader.LastHeader(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	newTipCap, err := p.client.SuggestGasTipCap(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	if lastTipCap != nil {
+		newTipCap = arbmath.BigMax(newTipCap, arbmath.BigMulByBips(lastTipCap, minRbfIncrease))
+	}
+	newFeeCap := new(big.Int).Mul(latestHeader.BaseFee, big.NewInt(2))
+	newFeeCap.Add(newFeeCap, newTipCap)
+
+	elapsed := time.Since(dataCreatedAt)
+	maxFeeCap := new(big.Int).SetUint64(uint64(p.config.MaxFeeCapGwei * params.GWei))
+	maxFeeCapDoublings := int64(elapsed / p.config.MaxFeeCapDoubling)
```

Comment:
Unfortunately we need it to be a step function to some degree with the replacement times, because geth requires that any increase in the fee cap must also be combined with a 10% increase in the tip cap. We don't want to keep increasing the tip cap as that is very rarely necessary, so we avoid increasing the fee too much.

Reply:
I suppose for the max fee cap, it doesn't need to be a step function though. We can just look at the current value when the next replace step is hit. I'll take a look at making this a smoother function.